### DISCUSSION TITLE
Error LengthAwarePaginator when trying to display posts via foreach and component.

### DISCUSSION THREAD

**Discussion by WillTheDeveloper** [livewire/livewire#3437] [1. Help]
### What's happening:

So recently I moved a lot of my blade component parts to livewire components and some of it works but other parts do not work as well as I hoped. I will note that because of the transition, everything is really messy. Only because livewire is really fast and way more efficient than some default Laravel stuff and it will allow me to really easily paginate stuff with specific requests to the database.

I have asked absolutely everywhere I can but its so confusing so this is my last resort. I have tried to give as much information as possible but feel free to ask for any code if you need it to help solve it.

**What it does at the moment:**
It errors something to do with "LengthAwarePagination", whatever that means. It is a mess because it uses a controller and livewire to grab the data but I am not sure how else I am supposed to get it to work.

![image](https://user-images.githubusercontent.com/78596837/127487797-a75f3e8e-80d3-478f-bcc0-75c8bdfc5442.png)

**What I want it to do:**
I want to be able to use a livewire component to display each of the posts with all the data inside of it with a foreach loop containing it. The way i think about it, if i can get it to work then the controllers should not be needed anymore and can be changed back to normal routes. Not sure how livewire handles magic methods but its a must otherwise none of it will work at all because the data is all over the place inside the database.

**What I need help with:**
Getting it to work the way it should with the livewire stuff. And the outcome should look like this (**This image uses controllers**):

![image](https://user-images.githubusercontent.com/78596837/127487716-d6159edb-417b-454f-878a-d23ef0cb2f35.png)

### The code:

**This is my discussion blade file, the page that all the posts should be visible on:**

```Blade
<div class="flex bg-white">
                <div class="hidden lg:block lg:col-span-3 xl:col-span-2">
                    <div class="flex flex-col w-44 sticky top-4">
                        <!-- Sidebar component, swap this element with another sidebar if you like -->
                        <div>
                            <livewire:forum-nav />
                        </div>
                    </div>
                </div>
                <div class="flex flex-col w-0 flex-1">

                    <livewire:forum-filters/>

                    <div class="pb-6">
                        <div class="max-w-7xl mx-auto px-4 sm:px-6 md:px-8">
                            <!-- Start Content -->
                            <div>
                                <ul class="space-y-8">
                                    @foreach($Posts as $Post)
                                        <livewire:view-post :Post="$Post" :wire:key="$Post->id"/>
                                    @endforeach
                                </ul>
                            </div>

                        <!-- End Content -->
                        </div>
                        {{$Posts->links()}}
                    </div>

                </div>
            </div>
```

**This is the ViewPost.php class file**

```php
use WithPagination;

    public function render()
    {
        /*$Posts = Posts::with(['Users', 'Categories', 'Comments'])
                    ->addSelect(['title', 'message', 'id', 'users_id'])
                    ->orderBy('created_at', 'desc')
                    ->with('Users')
                    ->withCount('Comments');
        dd($Posts);*/

        return view('livewire.view-post', [
            'Post' => Posts::with(['Users', 'Categories', 'Comments', 'Channels'])
                ->addSelect(['title', 'message', 'id', 'users_id'])
                ->orderBy('created_at', 'desc')
                ->withCount('Comments')
                ->paginate('10')
        ]);

    }
```

**This is my controller which will become redundant**

```php
public function get()
    {
        $Posts = Posts::all()->sortByDesc('id');
//        $Posts = Posts::with('Posts', 'Comments', 'Users', 'AccountTypes', 'Channels')->simplePaginate(5);
/*          $Comments = Comments::all();*/
//        $Comments = Comments::all()->where('posts_id', $this->comments()->id);

        return view('forum.discussion', compact(['Posts']));
    }
```

**This is the blade file for the livewire component which has all the links to the database table in. This is the template for each of the posts. Part of it anyway because it huge.**

```Blade
<div class="min-w-0 self-center flex-1">
                        <div class="flex items-center text-sm text-gray-400 space-x-2">
                            <div>
                                <p class="text-sm font-medium text-gray-900">
                                    <a href="/userprofile/{{$Post->users->id}}"
                                       class="hover:underline">{{$Post->users->name}}</a>
                                </p>
                            </div>
                            <div>&bull;</div>
                            <div>{{$Post->users->accounttype->accounttype}}</div>
                        </div>
                        <div>
                            <p class="text-sm text-gray-500">
                                <a href="#" class="hover:underline">
                                    <time>{{$Post->created_at->diffForHumans()}}</time>
                                </a>
                            </p>
                        </div>
                    </div>
```

**Comment by ph7jack** [comment]
Hello @WillTheDeveloper 

I recommend to always create a consistent wire:key, like `wire:key="posts.{{ $post->id }}"`

Your error is because the `$Post` variable on `ViewPost` component not is a eloquent post model, is a **paginator instance.**
You are passing in component attributes the `Post` variable, but this is not defined on **ViewPost** component, and you are passing a new Post variable returning a paginator instance
```
return view('livewire.view-post', [
            // This is the error, you don't need it
            'Post' => Posts::with(['Users', 'Categories', 'Comments', 'Channels'])
                ->addSelect(['title', 'message', 'id', 'users_id'])
                ->orderBy('created_at', 'desc')
                ->withCount('Comments')
                ->paginate('10')
        ]);
```

You refactored component

```
class ViewPost extends component {
    use WithPagination;

    public Post $post;

    public function render()
    {
        return view('livewire.view-post');
    }
}
```

Explanation:
You have a foreach with all posts 
```
@foreach($Posts as $Post)
    <livewire:view-post :Post="$Post" :wire:key="$Post->id"/>
@endforeach
```

Each row will create a new Livewire **ViewPost** component.
The **ViewPost** component will receives a `$post` variable (post model instance). In this, only need define a public property $post to use it

**Reply by WillTheDeveloper** [reply]
Okay, so this is what I have at the moment, following as much as what you said as possible, but it still give me an error.

![image](https://user-images.githubusercontent.com/78596837/127737396-c7e417cb-aac9-4b04-b84e-b1021cf06b6f.png)

Not sure what you wanted me to do with the foreach statement so I left it like this:

```blade
<div>
                                <ul class="space-y-8">
                                    @foreach($Posts as $Post)
                                        <livewire:view-post :Post="$Post" :wire:key="$Post->id"/>
                                    @endforeach
                                </ul>
                            </div>
```

Then on the ViewPost.php I have commented out everything else because you said that is what was giving me the error:

```php
class ViewPost extends Component
{
    use WithPagination;

    public Posts $post;

    public function render()
    {
        /*$Posts = Posts::with(['Users', 'Categories', 'Comments'])
                    ->addSelect(['title', 'message', 'id', 'users_id'])
                    ->orderBy('created_at', 'desc')
                    ->with('Users')
                    ->withCount('Comments');
        dd($Posts);*/

        return view('livewire.view-post', [
            /*'Post' => Posts::with(['Users', 'Categories', 'Comments', 'Channels'])
                ->addSelect(['title', 'message', 'id', 'users_id'])
                ->orderBy('created_at', 'desc')
                ->withCount('Comments')
                ->paginate('10')*/
        ]);

    }
}
```

I don't suppose I need to do anything with the controller? The controller is still getting the collection from the database.
The controller is the same snippet of code as my previous message.

I am not very good at livewire :(

**Reply by ph7jack** [reply]
Can you show the **NewPost** class?

**Reply by WillTheDeveloper** [reply]
NewPost.php is here:

```php
<?php

namespace App\Http\Livewire;

use App\Models\Comments;
use App\Models\Posts;

use App\Models\User;
use Livewire\Component;


class NewPost extends Component
{
    public Posts $posts;

    public $title;
    public $message;
    public $channels_id;


    protected $rules = ([
        'title' => 'required|string',
        'channels_id' => 'required|numeric',
        'message' => 'required|string',
    ]);

    /**
     * @var mixed
     */

    public function save(Posts $posts)
    {
        $this->validate();

        if (auth()->check()) {
            $posts->Posts()->create(
                [
                    'users_id' => Auth()->id(),
                    'channels_id' => $this->channels_id,
                    'title' => $this->title,
                    'message' => $this->message,
                ]
            );
        }

        return redirect('discussion');
    }

    public function render()
    {
        return view('livewire.new-post');
    }
}
```
